tests/robustness/model/model.go

```diff
@@ -33,469 +30,92 @@ const (
 	if len(states) == 0 {
 		// states were not initialized
 		if response.Err != nil || response.ResultUnknown || response.Revision == 0 {
 			return true, nil
 		}
-		return true, PossibleStates{initState(request, response)}
+		return true, initNonDeterministicState(request, response)
 	}
+	var newStates nonDeterministicState
 	if response.Err != nil {
-		states = applyFailedRequest(states, request)
+		newStates = states.stepFailedRequest(request)
 	} else {
-		states = applyRequest(states, request, response)
+		newStates = states.stepSuccessfulRequest(request, response)
 	}
-	return len(states) > 0, states
+	return len(newStates) > 0, newStates
 }
 
-// initState tries to create etcd state based on the first request.
-func initState(request EtcdRequest, response EtcdResponse) EtcdState {
-	state := EtcdState{
-		Revision:  response.Revision,
-		KeyValues: map[string]ValueRevision{},
-		KeyLeases: map[string]int64{},
-		Leases:    map[int64]EtcdLease{},
-	}
-	switch request.Type {
-	case Txn:
-		if response.Txn.TxnResult {
-			return state
-		}
-		if len(request.Txn.Ops) != len(response.Txn.OpsResult) {
-			panic(fmt.Sprintf("Incorrect request %s, response %+v", describeEtcdRequest(request), describeEtcdResponse(request, response)))
-		}
-		for i, op := range request.Txn.Ops {
-			opResp := response.Txn.OpsResult[i]
-			switch op.Type {
-			case Range:
-				for _, kv := range opResp.KVs {
-					state.KeyValues[kv.Key] = ValueRevision{
-						Value:       kv.Value,
-						ModRevision: kv.ModRevision,
-					}
-				}
-			case Put:
-				state.KeyValues[op.Key] = ValueRevision{
-					Value:       op.Value,
-					ModRevision: response.Revision,
-				}
-			case Delete:
-			default:
-				panic("Unknown operation")
-			}
-		}
-	case LeaseGrant:
-		lease := EtcdLease{
-			LeaseID: request.LeaseGrant.LeaseID,
-			Keys:    map[string]struct{}{},
-		}
-		state.Leases[request.LeaseGrant.LeaseID] = lease
-	case LeaseRevoke:
-	case Defragment:
-	default:
-		panic(fmt.Sprintf("Unknown request type: %v", request.Type))
-	}
-	return state
+func initNonDeterministicState(request EtcdRequest, response EtcdNonDeterministicResponse) nonDeterministicState {
+	return nonDeterministicState{initState(request, response.EtcdResponse)}
 }
 
-// applyFailedRequest handles a failed requests, one that it's not known if it was persisted or not.
-func applyFailedRequest(states PossibleStates, request EtcdRequest) PossibleStates {
+// stepFailedRequest duplicates number of states by considering request persisted and lost.
+func (states nonDeterministicState) stepFailedRequest(request EtcdRequest) nonDeterministicState {
+	newStates := make(nonDeterministicState, 0, len(states)*2)
 	for _, s := range states {
-		newState, _ := applyRequestToSingleState(s, request)
+		newStates = append(newStates, s)
+		newState, _ := s.step(request)
 		if !reflect.DeepEqual(newState, s) {
-			states = append(states, newState)
```

Comment:
nit: I see that the check was added in different PR, but can you add a comment why we need `!reflect.DeepEqual(newState, s)`

Reply:
I guess it is just used to avoid same states. Please confirm.

Reply:
If the request doesn't impact the state then there is no reason to fork it. Example is failed defrag request.